feature: add gzip support for static files

siamskoi opened the pull request:
# Reason for This PR

Static file compression functionality is required. Fileserver is not suitable for this, as it requires nginx or apache for its use. With nginx or apache, it makes no sense to use the RR functionality for working with static files.

## Description of Changes

The functionality of compressing static files using gzip has been added, the corresponding header is added to the response, a .gz file is created to reduce the load on the CPU for re-compressing files. Added a `gzip` setting to enable this functionality.

`curl -I -X GET --header "Accept-Encoding: gzip" http://127.0.0.1:8080/style.css`

Answer for small files:
> HTTP/1.1 200 OK
> Accept-Ranges: bytes
> Content-Encoding: gzip
> Content-Type: text/css; charset=utf-8
> Last-Modified: Tue, 28 Jan 2025 11:10:13 GMT
> Vary: Accept-Encoding
> Date: Tue, 28 Jan 2025 17:07:22 GMT
> **Content-Length: 46**

And answer for big files:
> HTTP/1.1 200 OK
> Accept-Ranges: bytes
> Content-Encoding: gzip
> Content-Type: text/css; charset=utf-8
> Last-Modified: Thu, 16 Jan 2025 20:09:56 GMT
> Vary: Accept-Encoding
> Date: Tue, 28 Jan 2025 17:32:31 GMT
> **Transfer-Encoding: chunked**


## License Acceptance

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [X] The reason for this PR is clearly provided (issue no. or explanation).
- [X] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [X] All added/changed functionality is tested.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Added support for optional gzip compression when serving static files
	- Implemented file compression mechanism for improved content delivery efficiency

- **Bug Fixes**
	- Enhanced error handling during file compression and serving processes

- **Performance**
	- Introduced thread-safe compression caching to optimize file serving

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

siamskoi committed:
Add gzip support for static files

rustatian commented on plugin.go:
```diff
@@ -23,6 +27,8 @@ const (
 	RootPluginName = "http"
 )
 
+var cacheMutex sync.Mutex
+
```
No global variables (this is a bad practice in Go) allowed

rustatian commented on plugin.go:
```diff
@@ -225,11 +234,78 @@ func (s *Plugin) Middleware(next http.Handler) http.Handler { //nolint:gocognit,
+			if err != nil {
+				s.log.Error("failed to close file", zap.Error(err))
+			}
+			f = cf
```
I'm not sure what is that. You're closing a file and then reassign another file to the closed one to be closed by the initial defer?

rustatian commented on plugin.go:
```diff
@@ -225,11 +234,78 @@ func (s *Plugin) Middleware(next http.Handler) http.Handler { //nolint:gocognit,
 }
 
+func (s *Plugin) createCompressedFile(originalFile http.File, compressedPath string) error {
+	cacheMutex.Lock()
```
No globals

rustatian commented on plugin.go:
```diff
@@ -48,6 +54,8 @@ type Plugin struct {
 	// opentelemetry
 	prop propagation.TextMapPropagator
+
+	gzip bool
```
option without explanation comment. Also I think this is a bad name for the option, because what a `gzip=true` should do in the `gzip` middleware?

rustatian commented on plugin.go:
```diff
@@ -225,11 +234,78 @@ func (s *Plugin) Middleware(next http.Handler) http.Handler { //nolint:gocognit,
+	if err != nil {
+		return err
+	}
+	defer func() {
```
Too many unnececarry defers. There're no errors returning by another function calls, to overload a stack with 2 defers in a row.

rustatian commented on plugin.go:
```diff
@@ -225,11 +234,78 @@ func (s *Plugin) Middleware(next http.Handler) http.Handler { //nolint:gocognit,
+	}()
+
+	gz := gzip.NewWriter(compressedFile)
+	defer func() {
```
second unnecessary defer

rustatian commented:
Hey @siamskoi 👋🏻 
Thank you for the PR 👍🏻 

A few additional notes:
1. Tests are missing.
2. PR with docs to the roadrunner-server/docs is also missing.

siamskoi committed:
Enable configurable gzip compression with file size limits

Introduce `gzipEnabled` and `gzipMaxFileSize` for fine-grained gzip compression settings. Added logic to skip gzip for already compressed content and enforce file size limits. Improved thread safety and path validation in file compression.

siamskoi committed:
Enable configurable gzip compression with file size limits

Introduce `gzipEnabled` and `gzipMaxFileSize` for fine-grained gzip compression settings. Added logic to skip gzip for already compressed content and enforce file size limits. Improved thread safety and path validation in file compression.

siamskoi committed:
Refactor gzip handling to use klauspost/compress library

Replaced custom gzip logic with `github.com/klauspost/compress` to simplify compression handling and improve maintainability. Removed unused attributes and methods related to custom gzip implementation, such as max file size and custom file compression logic. This change streamlines the code and leverages the capabilities of an external, well-maintained library.

rustatian commented on config.go:
```diff
@@ -31,6 +31,9 @@ type Config struct {
 	Response map[string]string `mapstructure:"response"`
+
+	// GzipEnabled determines if gzip compression is enabled for serving static files.
+	GzipEnabled bool `mapstructure:"gzip_enabled"`
```
I'm not sure this is a good name for the option. From my POV option name should be self-explanatory.
I think that we may remove that option at all and check the `gzip` header only.

rustatian commented on plugin.go:
```diff
@@ -225,6 +231,26 @@ func (s *Plugin) Middleware(next http.Handler) http.Handler { //nolint:gocognit,
+		if s.gzipEnabled && strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") {
+			s.log.Debug("gzip compression requested")
+			// Skip compression for already compressed formats
+			contentType := http.DetectContentType(make([]byte, 512))
```
Are you detecting the content-type for the empty slice of bytes?
